src/python/pants/cache/artifact.py

```diff
@@ -109,30 +108,9 @@ def collect(self, paths):
         self._relpaths.add(relpath)
 
   def extract(self):
-    try:
-      with open_tar(self._tarfile, 'r', errorlevel=2) as tarin:
-        # Note: We create all needed paths proactively, even though extractall() can do this for us.
-        # This is because we may be called concurrently on multiple artifacts that share directories,
-        # and there will be a race condition inside extractall(): task T1 A) sees that a directory
-        # doesn't exist and B) tries to create it. But in the gap between A) and B) task T2 creates
-        # the same directory, so T1 throws "File exists" in B).
-        # This actually happened, and was very hard to debug.
-        # Creating the paths here up front allows us to squelch that "File exists" error.
-        paths = []
-        dirs = set()
-        for tarinfo in tarin.getmembers():
-          paths.append(tarinfo.name)
-          if tarinfo.isdir():
-            dirs.add(tarinfo.name)
-          else:
-            dirs.add(os.path.dirname(tarinfo.name))
-        for d in dirs:
-          try:
-            os.makedirs(os.path.join(self._artifact_root, d))
-          except OSError as e:
-            if e.errno != errno.EEXIST:
-              raise
-        tarin.extractall(self._artifact_root)
-        self._relpaths.update(paths)
-    except tarfile.ReadError as e:
-      raise ArtifactError(str(e))
+    # Note(yic): unlike the python implementation before, now we do not update self._relpath
+    # after the extraction.
+    result = self.NATIVE_BINARY.decompress_tarball(self._tarfile.encode('utf-8'),
+                                                   self._artifact_root.encode('utf-8'))
+    if result.is_throw:
```

Comment:
Rather than doing this here, let's add a call to `raise_or_return` to the `def decompress_tarball` in `native.py` so that this just raises at the FFI boundary.

(We probably then want to wrap the exception we get in an `ArtifactError` here, but encapsulating `PyResult` in `native.py` is probably worthwhile)

Reply:
Good idea! Adjusted